massa-bootstrap/src/server.rs

```diff
@@ -14,363 +42,470 @@ use massa_signature::KeyPair;
+                            data_execution,
+                            version,
+                            consensus_command_sender,
+                            network_command_sender,
+                        )
+                    });
+
+                massa_trace!("bootstrap.session.started", {
+                    "active_count": Arc::strong_count(&bootstrap_sessions_counter) - 1
+                });
+            } else {
+                let _ = match  self.runtime.block_on(server.send_error("Bootstrap failed because the bootstrap server currently has no slots available.".to_string())) {
                             Err(_) => Err(std::io::Error::new(std::io::ErrorKind::TimedOut, "bootstrap error no available slots send timed out").into()),
                             Ok(Err(e)) => Err(e),
                             Ok(Ok(_)) => Ok(()),
                         };
-                        debug!("did not bootstrap {}: no available slots", remote_addr);
-                    }
-                }
+                debug!("did not bootstrap {}: no available slots", remote_addr);
             }
         }
-        // wait for bootstrap sessions to finish
-        while bootstrap_sessions.next().await.is_some() {}
-
+        // TODO: when they are synchrenous, consider if we want to signal updater and listener
+        //       here, or in the stop method.
+        // TODO: do we drop(self.runtime) here?
         Ok(())
     }
 
-    #[cfg(test)]
-    // TODO we didn't test whether the peer IP address is banned
-    async fn is_ip_allowed(
-        &self,
-        remote_addr: SocketAddr,
-        server: BootstrapServerBinder,
-        _whitelist: &Option<HashSet<IpAddr>>,
-        _blacklist: &Option<HashSet<IpAddr>>,
-    ) -> io::Result<(BootstrapServerBinder, SocketAddr)> {
-        Ok((server, remote_addr))
-    }
-
-    #[cfg(not(test))]
-    // whether the peer IP address is banned
-    async fn is_ip_allowed(
-        &self,
-        remote_addr: SocketAddr,
-        mut server: BootstrapServerBinder,
-        whitelist: &Option<HashSet<IpAddr>>,
-        blacklist: &Option<HashSet<IpAddr>>,
-    ) -> io::Result<(BootstrapServerBinder, SocketAddr)> {
-        let ip = normalize_ip(remote_addr.ip());
-        // whether the peer IP address is blacklisted
-        let not_allowed_msg = if let Some(ip_list) = &blacklist && ip_list.contains(&ip) {
-            massa_trace!("bootstrap.lib.run.select.accept.refuse_blacklisted", {"remote_addr": remote_addr});
-            Some(format!("IP {} is blacklisted", &ip))
-        // whether the peer IP address is not present in the whitelist
-        } else if let Some(ip_list) = &whitelist && !ip_list.contains(&ip){
-            massa_trace!("bootstrap.lib.run.select.accept.refuse_not_whitelisted", {"remote_addr": remote_addr});
-            Some(format!("A whitelist exists and the IP {} is not whitelisted", &ip))
+    /// this is a bit complicated, but the simple is bad:
```

Comment:
Why you don't use the select macro of crossbeam ?

Reply:
The select macro is non-deterministic in which branch it runs.

this approach allows us to wake up on the listen channel without actually consuming anything, and then checking for a stop signal.